### PR TITLE
Dev andrei

### PR DESCRIPTION
**Pull request by soldofor**
Baby steps

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -11,21 +11,19 @@ def validate(keyboard, pin):
     :param pin: a string of digits
     :return: nothing or raises ValueError if validation fails
     """
-    if not all([isinstance(keyboard, list), isinstance(pin, str)]):
+    if not all([isinstance(keyboard, list), isinstance(pin, str), keyboard != []]):
```
Comment (kopsha) [inline app.py]: Protecting against invalid input is the only way to survive in production. 👍🏾 

Suggestion: All validations worth to be separated and raise the correct exception, as follows:
- if pin is not string, `raise TypeError(msg)`
- if keyboard is not list, `raise TypeError(msg)`
- if keyboard is empty, `raise ValueError(msg)`

Comment (kopsha) [reply app.py]: Maybe we can reject an empty pin also.

File: tests.py [hunk 2]
```diff
@@ -9,7 +9,7 @@ def test_app():
         "2",
     ], f"Wrong result for {keyboard} and PIN = 1"
 
-    actual_value = pin_decoder(keyboard, "123")
+    actual_value = sorted(pin_decoder(keyboard, "123"))
```
Comment (kopsha) [inline tests.py]: To simplify the hard-coded values, let's choose a pin of length 1 and another one of length 2.

File: tests.py [hunk 3]
```diff
@@ -45,12 +45,12 @@ def test_app():
         "236",
         "232",
         "213",
-        "216",
         "212",
+        "216"
     ]
-    assert actual_value == expected_value, "Test case bla bla failed on empty pin"
-
+    assert actual_value == sorted(expected_value), "Test case failed on pin = 123"
```
Comment (kopsha) [inline tests.py]: As a rule of thumb, we should enunciate any error message in such way that we can change the values / parameters in the future. How about we completely remove the explanatory message, for now.

**Review by kopsha — CHANGES_REQUESTED** [review]
The baby steps are moving us towards the right direction.

File: app.py [hunk 4]
```diff
@@ -12,17 +12,17 @@ def validate(keyboard, pin):
...
         )
-    if not pin.isdigit():
-        raise ValueError(f"Input is not valid. Pin must be a string of digits.")
-    if not all([isinstance(el, list) for el in keyboard]) or not all(
-        [isinstance(el[i], str) for el in keyboard for i in range(len(el))]
-    ):
-        raise ValueError(
-            f"Input is not valid. Keyboard must be a list of lists of strings."
-        )
+    if keyboard == [] or pin == "":
```
Comment (kopsha) [inline app.py]: checking for emptiness is widely accepted as checking for *falsy* values, e.g. `if not keyboard: ...` or `if not pin: ...`

File: app.py [hunk 5]
```diff
@@ -12,17 +12,17 @@ def validate(keyboard, pin):
...
+        raise ValueError("Keyboard or PIN cannot be empty.")
```
Comment (kopsha) [inline app.py]: There is no added value to merge two errors in a single exception. At worst, when you will get this error (unexpectedly) you will wonder which is it the pin or the keyboard.

File: app.py [hunk 6]
```diff
@@ -12,17 +12,17 @@ def validate(keyboard, pin):
...
+    if not all([isinstance(row, list) for row in keyboard]):
+        raise TypeError("Keyboard must be a list of lists.")
+    if not all([isinstance(key, str) for row in keyboard for key in row]):
+        raise TypeError("Keyboard must be a list of lists of strings.")
+    if len(set(pin) - set(key for lst in keyboard for key in lst)) != 0:
```
Comment (kopsha) [inline app.py]: Checking for length of set is atypical, most likely you want to know if it's empty or not and an empty set is a "falsy" value. Have a look at this stackoverflow answer: https://stackoverflow.com/a/39984051/7373966

File: tests.py [hunk 7]
```diff
@@ -1,57 +1,16 @@
...
-        "233",
-        "236",
-        "232",
-        "213",
-        "216",
-        "212",
-    ]
-    assert actual_value == expected_value, "Test case bla bla failed on empty pin"
+def test_app():
+    test_pairs = [("13", ["22", "16", "12", "43", "46", "42", "23", "26", "13"]), ("1", ["2", "4", "1"]),
```
Comment (kopsha) [inline tests.py:7]: Nicely done, now we can add more test-cases just by extending this list.

**Review by kopsha — APPROVED** [review]
Alright, this looks much better. A lot of code has been simplified or at least moved from a horizontal arrangement towards a vertical one.

File: app.py (hunk 4 above)
Comment (soldofor) [reply app.py]: I usually do it like that `if not x: ...` but I wanted to bring up the problem for these types of cases at our next talk:
For this case if the keyboard input would be some other falsy value( 0, {}, "", None, etc) not an empty list. I wanted it to fall under the TypeError, and to only have ValueError for empty lists or strings.

File: app.py (hunk 5 above)
Comment (soldofor) [reply app.py]: Roger! Explicit is good :)

File: app.py (hunk 6 above)
Comment (soldofor) [reply app.py]: Thanks, acknowledged!

File: app.py (hunk 4 above)
Comment (kopsha) [reply app.py]: Fair point. What if you check the type before the value ?

Comment (soldofor) [reply app.py]: Yes, I will follow that order! :)

Comment (soldofor) [reply app.py]: Pun intended, hope it's not unprofessional for PRs

**Comment by soldofor** [comment]
Addressed the last issues from the review. Hope the last re-re-request for review wasn't pointless :)

**Review by kopsha — APPROVED** [review]
E tzutz!

**Pull request merged**